### PR TITLE
Convert classes to records

### PR DESCRIPTION
**Pull request by idegtiarenko**
This converts following classes into records:
* SnapshotsInProgress.ShardSnapshotStatus
* RestoreInProgress.ShardRestoreStatus
* RepositoryCleanupInProgress.Entry
* SnapshotDeletionsInProgress.Entry

### CODE REVIEW

**Review by original-brownbear — APPROVED** [review]
LGTM neat :)

File: server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java [hunk 1]
```diff
@@ -761,43 +687,43 @@ private static Entry createClone(
...
                 version,
                 source,
                 shardStatusByRepoShardId,
                 Map.of(),
+                failure,
                 false
             );
         }
 
-        private Entry(
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java:784]: IMO there was some value in hiding the constructor of this class. It has tricky invariants and we do not want callers to just construct arbitrary values. Could you add the constructor to forbidden APIs?

**Review by DaveCTurner — CHANGES_REQUESTED** [review]

**Comment by idegtiarenko** [comment]
I removed `SnapshotsInProgress.Entry` as `snapshotIndices` fields are excluded from equals/hashCode. I would like not to change that yet

**Review by DaveCTurner — APPROVED** [review]
LGTM.

I haven't looked at `SnapshotsInProgress.Entry` in detail but if I had to guess I'd say that it's an oversight that we don't consider all the fields in the equality check.

**Comment by idegtiarenko** [comment]
>I haven't looked at SnapshotsInProgress.Entry in detail but if I had to guess I'd say that it's an oversight that we don't consider all the fields in the equality check.

Possibly. However given this aspect and overall complexity of `SnapshotsInProgress.Entry` I would like to return to it separately

**Pull request merged**